### DISCUSSION TITLE
MONAI Label - Pathology

### DISCUSSION THREAD

**Discussion by SachidanandAlle** [Project-MONAI/MONAILabel#599] [Ideas]
Semantic segmentation for WSI Images
--------------------------------------------

1. LOAD Image in Viewer (DigitalSlideArchive/QuPath) and select ROI
![image](https://user-images.githubusercontent.com/7339051/151807001-62455088-cce2-4a42-acd7-4e5c5a8d8760.png)

2. Annotate ROI (e.g. Tumor, Stroma etc..)
![image](https://user-images.githubusercontent.com/7339051/151807133-e62d4255-dfb3-4fae-a224-6efe089e5a2f.png)

3. Submit ROI to MONAI Label (WSI/ROI + Annotations) for Training
4. Train/Finetune a model
5. Run Inference on WSI and produce/show label masks
6. Repeat steps 1 to 5 on same WSI image for different ROI or try other WSI image

**Comment by andreped** [comment]
@SachidanandAlle In my PhD I am working with digital pathology, and I just recently published a paper on this exact topic (see [here](https://www.frontiersin.org/articles/10.3389/fmed.2021.816281/full)). However, our pipeline consisted of using a mixture of open solutions instead of integrating everything into a single platform. Having everything into a single platform would be more ideal, but it is very challenging and tedious to do so. Therefore we took advantage of the solutions that "did it best", and made it possible to communicate and transfer files between solutions easily.

Nonetheless, packing this into M-L could be very relevant for our research group (and the digital pathology field as well), as we could very easily integrate our architecture and pytorch code into the M-L framework.

Therefore, I am **very** positive to the idea! Just remember that AFAIK Slicer has very little (or no) support for rendering WSIs. For reading these images we often use [OpenSlide](https://github.com/openslide/openslide-python). Also, if you wish to annotate, you would need very good tools to do so, especially if you wish to have similar annotation user experience as with [QuPath](https://github.com/qupath/qupath). Could be quite a lot of work... If you wish to render images in Python, you could use [pyFAST](https://pypi.org/project/pyFAST/), which we are developing at SINTEF. Could be an idea to start from there and build upon pyFAST instead of starting from scratch.

There are also other people working on having a one-platform-does-all solution for active learning-based digital pathology (e.g., [QuickAnnotator](https://github.com/choosehappy/QuickAnnotator)). Would recommend checking how they did it :]

**Comment by SachidanandAlle** [comment]
For early access, please find the sample pathology app for MONAI Label under:
https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/pathology

> NOTE:: Things are in progress/under-development..

**Comment by SachidanandAlle** [comment]
Need some developers from QuPath to add a plugin over here:
https://github.com/Project-MONAI/MONAILabel/tree/main/plugins

The prototype is here.. (which is kind of simple reference implementation to access MONAI Label APIs for Infer)
https://github.com/SachidanandAlle/qupath/tree/main/qupath-extension-monailabel
(Few things are still kind of hack.. like enable ROI for MONAI Label using Ctrl+Click)

**Reply by SenhorUnk** [reply]
What do you mean by "hack.. like enable ROI for MONAI Label using Ctrl+Click"? 

I am a bit stuck with QuPath and MONAI Label. Installation of the MONAI plugin worked well, server is connected, submitting my annotations to the MONAI Label server works fine too. But running one of the 3 annotation models leads to the following error. (Detailed error in the terminal below):
![image](https://user-images.githubusercontent.com/111501065/194246256-0d8983cd-a6ea-48c7-8c09-3e0e684b9566.png)

The error shown in the terminal seems to be the same as discussed here: https://github.com/btcrabb/SlideSeg/issues/8
I tried different images and file formats but still facing the same problem. 

"[2022-10-06 09:24:00,412] [17916] [MainThread] [ERROR] (uvicorn.error:369) - Exception in ASGI application
Traceback (most recent call last):
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\fastapi\applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\middleware\cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\exceptions.py", line 93, in __call__
    raise exc
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 21, in __call__
    raise e
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\starlette\routing.py", line 65, in app
    response = await func(request)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\fastapi\routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\fastapi\routing.py", line 160, in run_endpoint_function
    return await dependant.call(**values)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\monailabel\endpoints\wsi_infer.py", line 113, in api_run_wsi_inference
    return run_wsi_inference(background_tasks, model, image, session_id, wsi, output)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\monailabel\endpoints\wsi_infer.py", line 97, in run_wsi_inference
    result = instance.infer_wsi(request)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\monailabel\interfaces\app.py", line 658, in infer_wsi
    infer_tasks = create_infer_wsi_tasks(request, image)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\monailabel\interfaces\utils\wsi.py", line 43, in create_infer_wsi_tasks
    with openslide.OpenSlide(image) as slide:
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\openslide\__init__.py", line 160, in __init__
    self._osr = lowlevel.open(filename)
  File "C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\lib\site-packages\openslide\lowlevel.py", line 128, in _check_open
    raise OpenSlideUnsupportedFormatError(
openslide.lowlevel.OpenSlideUnsupportedFormatError: Unsupported or missing image file"

**Reply by SachidanandAlle** [reply]
I see you are running on windows.. so you need OpenSlide lib installed.. Due to licensing, Monailabel can't help/install it.. user has to take care..

Steps are mentioned here..

https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/pathology#usage-development-mode

**Reply by SachidanandAlle** [reply]
Also if you are using GitHub version for monailabel (developer mode) you can try latest alpha version of qupath.. in that case you don't need images to be shared.. rather you can do Annotations on your own image/ROI and you can submit annotations for multiple ROIs per image..

**Reply by SenhorUnk** [reply]
Thanks for your response. I already installed openslide lib. According to the manual (below) I used this command to set the path: 
"set Path=%Path%;C:\UserData\z004muey\Documents\04_VS_Code_Workspace\MONAI_Label\openslide-win64-20220811\bin"

"Install [Openslide](https://openslide.org/) binaries manually and make sure .dll or .so files for openslide are in system load path.

For windows, make sure <openslide_folder>/bin is added in PATH environment."

Regarding your second comment: I was abel to submit annotations to the server already. But somehow I can't make use of the pretrained models since MONAI Label Server throws an error.

**Reply by SachidanandAlle** [reply]
Still getting openslide error?

**Reply by SenhorUnk** [reply]
I figured out, that openslide only supports a specific image format. I was using some sample .png, and .jpg, converted them to .tiff and still got the same error. I found a .svs file online and tried it too. The error changed to: 
![image](https://user-images.githubusercontent.com/111501065/194398495-bad71829-b285-4ae6-95ba-a6b2a03541fe.png)

The error appears to be related to openslide. Do you by any chance know, how to not use openslide on the MONAI server side, since my input images come as normal image. Thanks already for your help!

**Reply by SachidanandAlle** [reply]
You can use normal png images.. use the latest alpha version.. and use monailabel from main branch (developer mode.. uninstalled pip package).. and u don't need to include your images in the studies folder

**Reply by SenhorUnk** [reply]
Alpha Version of QuPath: 0.3.2 ?

I will try as suggested - thank you!

**Reply by SenhorUnk** [reply]
It worked :) thank you!

**Comment by andreped** [comment]
Might be that @petebankhead is interested in contributing.

Would be a great extension to QuPath and pathologists would surely embrace what this add-on has to offer. I also think it is a smart option to integrate digital pathology-related stuff into QuPath rather than Slicer, as pathologists are more familiar with QuPath. That everything is built upon MONAI also secures that new technology, architectures, and concepts/designs can easily be added and improved over time. Great idea, @SachidanandAlle !

**Comment by petebankhead** [comment]
Looks interesting!

I'm the only QuPath developer these days (but recruiting soon...) and completely swamped with tasks, so I'm afraid I can't do anything much directly right now.

However, I'm putting together plans for a webinar and hackathon starting (probably) 25 April*. It could be good to discuss further during that week, and perhaps work on any necessary extension then. By that time, I hope to have a new release, better developer documentation, and various enhancements available that could help.

*-No official announcement yet, but I plan to post about it on the [website](http://qupath.github.io), [twitter](https://twitter.com/qupath) and [image.sc](https://forum.image.sc/tag/qupath) within the next couple of weeks.

**Comment by SachidanandAlle** [comment]
@andreped have you tried the latest 0.4 release candidate and pathology example.. either on DSA or QuPath..?

Until we wait for QuPath developers to contribute and make it better, I could just give simple plugin to do basic things like annotate, submit and retrain the model.. of-course interaction can be made better.. specially in case of click interactions (and reduce mouse clicks to run the model)

**Comment by andreped** [comment]
@SachidanandAlle I have been quite occupied developing my own digital pathology software(s) recently :P, which just now had its first major release, [FastPathology](https://github.com/AICAN-Research/FAST-Pathology).

However, FastPathology is purely a deployment solution, and does not, nor intend to, support training and development of new models, as MONAI Label is meant for.

I could test MONAI Label with QuPath, if you'd like. I also have a pathologist who will definitely be interested in testing the plugin and providing feedback. @sahpet: Do you have time and are interesting in testing the plugin? We could do an interation together, if you'd like, @sahpet?
